### PR TITLE
Move generic code to strategies library.

### PR DESCRIPTION
**Pull request by sergey-shandar**
## Description

Depends on #5933

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [ ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+namespace Microsoft.Azure.Commands.Common.Strategies
+{
+    public static class AsyncCmdletExtensions
+    {
+        public static void WriteVerbose(this IAsyncCmdlet cmdlet, string message, params object[] p)
+            => cmdlet.WriteVerbose(string.Format(message, p));
+
+        public static string UpdateLocation(
+            this IState current, string location, IResourceConfig config)
+            => location ?? current.GetLocation(config) ?? "eastus";
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs]: do we always want to default to eastus for all usages of the strategies library?

File: src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            this IClient client,
+            string subscriptionId,
+            IParameters<TModel> parameters,
+            IAsyncCmdlet asyncCmdlet,
+            CancellationToken cancellationToken)
+            where TModel : class
+        {
+            // create a DAG of configs.
+            var config = await parameters.CreateConfigAsync();
+
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs]: remove extra vertical space between lines (only vertical space after ending ```}```

File: src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            if (s != null)
+            {
+                return "\"" + s + "\"";
+            }
+            var e = value as IEnumerable;
+            if (e != null)
+            {
+                return string.Join(",", e.Cast<object>().Select(ToPowerShellString));
+            }
+            return value.ToString();
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs]: needs vertical space after }

File: src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                        var progress = 0.0;
+                        var activeTasks = new List<string>();
+                        foreach (var taskProgress in asyncCmdlet.TaskProgressList)
+                        {
+                            if (!taskProgress.IsDone)
+                            {
+                                var config = taskProgress.Config;
+                                activeTasks.Add(config.GetFullName());
+                            }
+                            progress += taskProgress.GetProgress();
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs]: vertical space after '}' (please apply throughout file)

File: src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Commands.Common.Strategies
+{
+    public static class AsyncCmdletExtensions
+    {
+        public static void WriteVerbose(this IAsyncCmdlet cmdlet, string message, params object[] p)
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs]: Please provide documentation comments for this class, and for all public properties and methoids (you have most of the methods)

File: src/ResourceManager/Common/Commands.Common.Strategies/IAsyncCmdlet.cs [hunk 6]
```diff
@@ -13,18 +13,23 @@
...
 
 using Microsoft.Azure.Commands.Common.Strategies;
+using System.Collections.Generic;
 using System.Threading.Tasks;
 
-namespace Microsoft.Azure.Commands.Compute.Strategies
+namespace Microsoft.Azure.Commands.Common.Strategies
 {
-    interface IAsyncCmdlet
+    public interface IAsyncCmdlet
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/IAsyncCmdlet.cs]: Please provide documentation comments for this public interface and each of its methods and properties.

File: src/ResourceManager/Common/Commands.Common.Strategies/ICmdlet.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Commands.Common.Strategies
+{
+    public interface ICmdlet
+    {
+        void WriteVerbose(string message);
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/ICmdlet.cs]: WriteVerbose, WriteObject, and WriteProgress are already in ICommandRuntime.  Not sure why we need this - seems like just for Parameters.

Also, please provide documentation comemnts on the interface and properties.

File: src/ResourceManager/Common/Commands.Common.Strategies/IParameters.cs [hunk 8]
```diff
@@ -12,24 +12,15 @@
...
 
-using Microsoft.Azure.Commands.Common.Strategies;
-using System.Management.Automation;
 using System.Threading.Tasks;
 
-namespace Microsoft.Azure.Commands.Compute.Strategies
+namespace Microsoft.Azure.Commands.Common.Strategies
 {
-    internal sealed class ShouldProcess : IShouldProcess
+    public interface IParameters<TModel>
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/IParameters.cs]: Can we retroactively add documentation comments here?

File: src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs (hunk 1 above)
Comment (sergey-shandar) [reply src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs]: I'm not sure. We need to discuss this.

File: src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+                client,
+                target,
+                cancellationToken,
+                new ShouldProcess(asyncCmdlet),
+                asyncCmdlet.ReportTaskProgress);
+            // return a resource model
+            return newState.Get(config) ?? current.Get(config);
+        }
+
+        static string UpdateLocation(
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs]: I don't think that the default should be east us, I think this is something that should be set for each client (perhaps a DefaultLocation property.

The fear is that services that are not available in eastus will default to this value inadvertently

**Pull request merged**